src/c_arc.rs

```diff
@@ -0,0 +1,122 @@
+use crate::src::c_box::CBox;
+use crate::src::error::Rav1dResult;
+use std::marker::PhantomData;
+use std::ops::AddAssign;
+use std::ops::Deref;
+use std::ptr::NonNull;
+use std::sync::Arc;
+
+/// A C/custom [`Arc`].
+///
+/// That is, it is analogous to an [`Arc`],
+/// but it lets you set a C-style `free` `fn` for deallocation
+/// instead of the normal [`Box`] (de)allocator.
+/// It can also store a normal [`Box`] as well.
+///
+/// It is built around the [`CBox`] abstraction.
+/// However, that necessitates a double indirection
+/// to reach the ptr through the [`Arc`] and [`CBox`].
+/// To remedy this and improve performance,
+/// a stable pointer is stored inline,
+/// removing the double indirection.
+/// This self-referential ptr is sound
+/// because [`Arc`] and [`Box`]/[`CBox`]
+/// never move their data ptr during moves, as it's on the heap.
+/// As long as [`Self::owner`] is never moved
+/// without also re-updating [`Self::stable_ref`], this is sound.
+///
+/// Furthermore, storing this stable ref ptr like this
+/// allows for provenance projections of [`Self::stable_ref`],
+/// such as slicing it for a `CArc<[T]>`.
+#[derive(Debug)]
+pub struct CArc<T: ?Sized> {
+    owner: Arc<CBox<T>>,
```

Comment:
Could this be an `Arc<Pin<CBox<T>>>`? Should it? I think it can be and that would make the assumptions and requirements we're putting on the box explicit. Arc impls Unpin, so it wouldn't change where we can use CArc.

Reply:
I'm not too familiar with `Pin` so I wasn't sure.  I'll try it out and see if it works nicely; it'd be nice to move the unsafely and invariants into `Pin`.

Reply:
I `Pin`ned it in b0b6cdf.  Is that worth it?  It doesn't really improve the `unsafe`ty, I don't think, though it makes the safety invariants a little bit clearer.